### PR TITLE
Added BitSet Extract and Mask functions

### PR DESCRIPTION
**Pull request by davidtaylor-juul**

### CODE REVIEW

**Review by cedrickcooke — APPROVED** [review]
❤️

File: encoding/src/commonMain/kotlin/BitSet.kt [hunk 1]
```diff
@@ -4,6 +4,9 @@ public interface BitSet<T> {
     public operator fun get(index: Int): Boolean
     public operator fun set(index: Int, value: Boolean): BitSet<T>
     public fun asPrimitive(): T
+
+    /** Extracts [count] bits at [offset]. e.g. `0b00001100.extract(2, 2) = 0x03`. */
```
Comment (twyatt) [inline encoding/src/commonMain/kotlin/BitSet.kt:8]: nit

```suggestion
    /** Extracts [count] bits at [offset]. e.g. `0b00001100.extract(2, 2) = 0b11 = 0x03`. */
```

File: encoding/src/commonMain/kotlin/BitSet.kt [hunk 2]
```diff
@@ -27,6 +30,12 @@ public data class IntBitSet(private var buffer: Int) : BitSet<Int> {
     }
 
     override fun asPrimitive(): Int = buffer
+
+    override fun extract(offset: Int, count: Int): Int {
+        require(offset in 0 until 32)
+        require(count in 0 until 32 - offset + 1)
```
Comment (twyatt) [inline encoding/src/commonMain/kotlin/BitSet.kt:36]: nit: Can we give these failure messages (that includes the value that was provided), please?

File: encoding/src/commonMain/kotlin/BitSet.kt [hunk 3]
```diff
@@ -45,4 +54,22 @@ public data class LongBitSet(private var buffer: Long) : BitSet<Long> {
     }
 
     override fun asPrimitive(): Long = buffer
+
+    override fun extract(offset: Int, count: Int): Long {
+        require(offset in 0 until 64)
+        require(count in 0 until 64 - offset + 1)
```
Comment (twyatt) [inline encoding/src/commonMain/kotlin/BitSet.kt:60]: Here too.

File: encoding/src/commonMain/kotlin/BitSet.kt [hunk 4]
```diff
@@ -45,4 +54,22 @@ public data class LongBitSet(private var buffer: Long) : BitSet<Long> {
...
+        return (buffer ushr offset) and flatLongBitMask(count)
+    }
+}
+
+/** Create an [Int] mask of `count` length e.g. flatBitMask(3) = 0b00000111`. */
+public fun flatBitMask(count: Int): Int {
+    require(count in 0..32)
```
Comment (twyatt) [inline encoding/src/commonMain/kotlin/BitSet.kt:67]: Here too.

File: encoding/src/commonMain/kotlin/BitSet.kt [hunk 5]
```diff
@@ -45,4 +54,22 @@ public data class LongBitSet(private var buffer: Long) : BitSet<Long> {
...
+    return if (count == 32) -1 else (1 shl count) - 1
+}
+
+/** Create a [Long] mask of `count` length e.g. flatLongBitMask(3) = 0b00000111L`. */
+public fun flatLongBitMask(count: Int): Long {
+    require(count in 0..64)
```
Comment (twyatt) [inline encoding/src/commonMain/kotlin/BitSet.kt:73]: Here too (last one I think).

**Review by twyatt — APPROVED** [review]
🚀

File: encoding/src/commonMain/kotlin/BitSet.kt [hunk 6]
```diff
@@ -27,6 +30,12 @@ public data class IntBitSet(private var buffer: Int) : BitSet<Int> {
     }
 
     override fun asPrimitive(): Int = buffer
+
+    override fun extract(offset: Int, count: Int): Int {
```
[the remark this answers is not in the document]
Comment (twyatt) [reply encoding/src/commonMain/kotlin/BitSet.kt:34]: Feel free to disregard this comment if you feel it is more noise on the API than helpful.

File: encoding/src/commonMain/kotlin/BitSet.kt [hunk 7]
```diff
@@ -4,8 +4,15 @@ public interface BitSet<T> {
     public operator fun get(index: Int): Boolean
     public operator fun set(index: Int, value: Boolean): BitSet<T>
     public fun asPrimitive(): T
+
+    /** Extracts [count] bits at [offset]. e.g. `0b00001100.extract(2, 2) = 0b11 = 0x03`. */
+    public fun extract(offset: Int, count: Int): T
 }
 
+/** Extracts [range] bits . e.g. `0b00001100.extract(2..3) = 0b11 = 0x03`. */
```
Comment (twyatt) [inline encoding/src/commonMain/kotlin/BitSet.kt:12]: nit: extra whitespace

```suggestion
/** Extracts [range] bits. e.g. `0b00001100.extract(2..3) = 0b11 = 0x03`. */
```
```suggestion
/** Extracts [range] bits . e.g. `0b00001100.extract(2..3) = 0b11 = 0x03`. */
```

**Pull request merged**